vignettes/creating_linters.Rmd

```diff
@@ -11,113 +11,186 @@ vignette: >
-closure that will do the actual linting. We could define additional parameters
-that are useful for the linter in this function declaration, but
-`assignment_linter` requires no additional arguments.
+closure that will do the actual linting (i.e., they're function factories). We could define additional parameters
+that are useful for the linter in this function declaration (see, e.g. `assignment_linter`), but
+`pipe_call_linter` requires no additional arguments.
 
 ```r
-Linter(function(source_file) {
+Linter(function(source_expression) {
 ```
-This is the closure. It will be called on the `source_file` variable that
+
+This is the closure. It will be called on the `source_expression` variable that
 contains the top level expressions in the file to be linted. The call to
-`Linter()` simply gives this closure the class 'linter'.
+`Linter()` gives this closure the class 'linter' and guesses its name.
+
+The raw text of the expression is available from `source_file$content`. However,
+it is not generally possible to implement linters from the raw text -- consider
+`equals_na_linter`. If we just look for `== NA` in the text of the file, we'll
+generate many false positives, e.g. in comments (such as `# note: is.na() is the proper way to check == NA`)
+or inside character literals (such as `warning("don't use == NA to check missingness")`).
+We're also likely to generate false negatives, for example when `==` and `NA` appear on different lines!
+Working around these issues using only the un-parsed text in every situation amounts to re-implementing the parser.
 
-The raw text of the expression is available from `source_file$content`. However
-it is recommended to work with the tokens from `source_file$parsed_content` if
-possible, as they are tokenized from the `R` parser. These tokens are obtained
-from `parse()` and `getParseData()` calls done prior to calling the new linter.
+Therefore it is recommended to work with the tokens from `source_file$parsed_content` or `source_file$xml_parsed_content`,
+as they are tokenized from the `R` parser. These tokens are obtained
+from `parse()` and `utils::getParseData()` calls done prior to calling the new linter.
 `getParseData()` returns a `data.frame` with information from the source parse
 tree of the file being linted. A list of tokens is available from
-[r-source/src/main/gram.y](https://github.com/wch/r-source/blob/ff1bca2f21aba271d428474f00494eece5c36dd3/src/main/gram.y#L293-L307).
+[r-source/src/main/gram.y](https://github.com/r-devel/r-svn/blob/master/src/main/gram.y#L395-L412).
+
+`source_file$xml_parsed_content` uses `xmlparsedata::xml_parse_data()` to convert the `getParseData()` output into an XML tree,
+which enables writing linter logic in [XPath](https://www.w3schools.com/xml/xpath_intro.asp), a powerful language
+for expressing paths within the nested XML data structure. Most linters in `lintr` are built using XPath because it is
```

Comment:
WDYT about "...a powerful language for computation on the abstract syntax tree / parse tree"?